daphne/src/taskprov.rs

```diff
@@ -0,0 +1,171 @@
+/// specified pseudorandom key.
+pub(crate) fn compute_vdaf_verify_key_from_prk(
+    prk: &Prk,
+    task_id: &Id,
+    vdaf_type: VdafType,
+) -> VdafVerifyKey {
+    let info = [task_id.as_ref()];
+    match &vdaf_type {
+        VdafType::Prio3Aes128Count | VdafType::Prio3Aes128Sum | VdafType::Prio3Aes128Histogram => {
+            // This expand(), and the associated fill() below can only fail if the length is wrong,
+            // and it won't be, so we unwrap().
+            let okm = prk.expand(&info, vdaf_type.clone()).unwrap();
+            let mut bytes = [0u8; 16];
+            okm.fill(&mut bytes[..]).unwrap();
+            VdafVerifyKey::Prio3(bytes)
+        }
+        _ => panic!("Unknown VDAF type"),
+    }
+}
+
+/// Compute the VDAF verify key for task_id and the specified VDAF type using the
+/// pre-shared secret verify_key_init.
+///
+/// This is a convenience function to call compute_vdaf_verify_prk() and
+/// compute_vdaf_verify_key_from_prk().  Callers reusing the same PRK frequently
+/// should consider computing the prk once and then calling compute_vdaf_verify_key_from_prk()
+/// directly.
+#[allow(dead_code)]
+pub(crate) fn compute_vdaf_verify_key(
+    verify_key_init: &[u8],
+    task_id: &Id,
+    vdaf_type: VdafType,
+) -> VdafVerifyKey {
+    compute_vdaf_verify_key_from_prk(
+        &compute_vdaf_verify_prk(verify_key_init),
+        task_id,
+        vdaf_type,
+    )
+}
+
+pub fn bad_request(detail: &str) -> DapError {
+    DapError::Abort(DapAbort::BadRequest(detail.to_string()))
+}
+
+/// Check for a taskprov extension in the report, and return it if found.
+pub fn get_taskprov_task_config(
+    task_id: &Id,
+    metadata: &ReportMetadata,
+) -> Result<Option<TaskConfig>, DapError> {
+    let taskprovs: Vec<&Extension> = metadata
+        .extensions
+        .iter()
+        .filter(|x| matches!(x, Extension::Taskprov { .. }))
+        .collect();
+    match taskprovs.len() {
+        0 => Ok(None),
+        1 => match &taskprovs[0] {
+            Extension::Taskprov { payload } => {
+                if compute_task_id(&payload[..]) != *task_id {
+                    return Err(bad_request(
+                        "taskprov task config computed task id does not match report task id",
+                    ));
+                }
+                let task_config = TaskConfig::get_decoded(payload)
+                    .map_err(|_| bad_request("could not parse taskprov task config"))?;
+                Ok(Some(task_config))
+            }
+            _ => panic!("cannot happen"),
+        },
+        _ => Err(DapError::Abort(DapAbort::BadRequest(
+            "more than one taskprov extension".to_string(),
+        ))),
+    }
+}
+
+fn url_from_bytes(bytes: &[u8]) -> Result<Url, DapError> {
+    let s = str::from_utf8(bytes).map_err(|_| bad_request("bad URL UTF8"))?;
+    Url::parse(s).map_err(|_| bad_request("bad URL syntax"))
```

Comment:
Ah yes! We definitely need to handle this case by aborting, but we want to make sure we properly ["opt-out"](https://datatracker.ietf.org/doc/html/draft-wang-ppm-dap-taskprov-00#section-3.3). The protocol behavior depends somewhat the situation, but should be spelled out in the Leader Behavior and Helper Behavior sections of the taskprov spec.

Reply:
Will work on this.